src/core/lib/event_engine/posix.h

```diff
@@ -0,0 +1,159 @@
+// Copyright 2022 gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_POSIX_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_POSIX_H
+
+#include <grpc/support/port_platform.h>
+
+#include <functional>
+#include <vector>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+
+#include <grpc/event_engine/endpoint_config.h>
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/memory_allocator.h>
+#include <grpc/event_engine/port.h>
+#include <grpc/event_engine/slice_buffer.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+#ifdef GRPC_EVENT_ENGINE_POSIX
+
+/// This defines an EventEngine interface that posix specific event engines
+/// may implement to support additional functionality.
+class PosixEngine : public EventEngine {
+ public:
+  class PosixEventEngineEndpoint : public EventEngine::Endpoint {
```

Comment:
The names are going to be confusing, for instance:

```
class PosixEventEngine : public PosixEngine {
  class Endpoint : public PosixEngine::PosixEventEngineEndpoint {
    ...
  }
}
```

Using these with an interface like `void DoAThing(PosixEngine::PosixEventEngineEndpoint* endpoint) { ...  }`, you're probably going to have to look it up to determine if it the argument is taking the interface or the implementation (`PosixEventEngine::Endpoint` is the implementation).

I'd like to suggest an attribute-based alternative for these interfaces, not extending the base types, including strictly the added functionality. For instance:

```
// an Endpoint interface
class EndpointFdSupport {
  virtual int GetWrappedFd() = 0;
  /* ... */
};

class ListenerFdSupport {
  virtual absl::StatusOr<int> BindWithFd(
      const ResolvedAddress& addr,
      OnPosixBindNewFdCallback on_bind_new_fd) = 0;  
  /* ... */
};

class EventEngineFdSupport {
  virtual std::unique_ptr<PosixEventEngineEndpoint> CreatePosixEndpointFromFd(
      int fd, const EndpointConfig& config,
      MemoryAllocator memory_allocator) = 0;  
  /* ... */
}
```

What we'll have then is something like:

```
class PosixEventEngine : public EventEngine, public EventEngineFdSupport {
  class Endpoint : public EndpointFdSupport {};
};
```

And usage would look like `void DoAThing(EndpointFdSupport* endpoint) { ... }`

Reply:
Maybe `EndpointWithFdSupport` describes the object a bit better.

@ctiller does this mesh with your ideas here as well?